### PR TITLE
Use more range-based for loops.

### PR DESCRIPTION
**Pull request by masterleinad**
Some `clang-tidy` findings with `modernize-loop-convert` (and some manual adjustments).

### CODE REVIEW

File: include/deal.II/differentiation/ad/ad_helpers.h [hunk 1]
```diff
@@ -3794,10 +3794,10 @@ namespace Differentiation
 #    ifdef DEBUG
       const std::vector<unsigned int> index_set(
         internal::extract_field_component_indices<dim>(extractor));
-      for (unsigned int i = 0; i < index_set.size(); ++i)
+      for (unsigned int index : index_set)
```
Comment (bangerth) [inline include/deal.II/differentiation/ad/ad_helpers.h:3797]: can you make it `const`?

File: include/deal.II/matrix_free/face_setup_internal.h [hunk 2]
```diff
@@ -738,9 +738,9 @@ namespace internal
                                     cells_close_to_boundary.end());
       std::vector<unsigned int> final_cells;
       final_cells.reserve(cells_close_to_boundary.size());
-      for (unsigned int i = 0; i < cells_close_to_boundary.size(); ++i)
-        if (at_processor_boundary[cells_close_to_boundary[i]] == false)
-          final_cells.push_back(cells_close_to_boundary[i]);
+      for (unsigned int cell : cells_close_to_boundary)
```
Comment (bangerth) [inline include/deal.II/matrix_free/face_setup_internal.h:741]:
```suggestion
      for (const unsigned int cell : cells_close_to_boundary)
```

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 3]
```diff
@@ -1190,8 +1190,8 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
         additional_data.cell_vectorization_categories_strict ||
         dof_handlers.active_dof_handler == DoFHandlers::hp;
       unsigned int dofs_per_cell = 0;
-      for (unsigned int no = 0; no < dof_info.size(); ++no)
-        dofs_per_cell = std::max(dofs_per_cell, dof_info[no].dofs_per_cell[0]);
+      for (auto &info : dof_info)
```
Comment (bangerth) [inline include/deal.II/matrix_free/matrix_free.templates.h:1193]:
```suggestion
      for (const auto &info : dof_info)
```

File: include/deal.II/matrix_free/operators.h [hunk 4]
```diff
@@ -1101,8 +1101,8 @@ namespace MatrixFreeOperators
     Assert(data.get() != nullptr, ExcNotInitialized());
     typename Base<dim, VectorType, VectorizedArrayType>::size_type total_size =
       0;
-    for (unsigned int i = 0; i < selected_rows.size(); ++i)
-      total_size += data->get_vector_partitioner(selected_rows[i])->size();
+    for (unsigned int selected_row : selected_rows)
```
Comment (bangerth) [inline include/deal.II/matrix_free/operators.h:1104]: same here and below

Comment (masterleinad) [reply include/deal.II/matrix_free/operators.h]: Fixed.

**Comment by masterleinad** [comment]
The failing (Actions) OSX serial test looks spurious...

**Pull request merged**